src/journal/bsod.c

```diff
@@ -0,0 +1,266 @@
+        _cleanup_free_ char *message = NULL;
+        const void *d;
+        size_t l;
+        sd_id128_t boot_id;
+        char boot_id_filter[STRLEN("_BOOT_ID=") + SD_ID128_STRING_MAX];
+        int r;
+
+        assert(ret);
+
+        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open journal: %m");
+
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get boot ID: %m");
+
+        xsprintf(boot_id_filter, "_BOOT_ID=" SD_ID128_FORMAT_STR, SD_ID128_FORMAT_VAL(boot_id));
+
+        r = sd_journal_add_match(j, boot_id_filter, 0);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to add boot ID filter: %m");
+
+        r = sd_journal_add_match(j, "_UID=0", 0);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to add User ID filter: %m");
+
+        assert_cc(0 == LOG_EMERG);
+        r = sd_journal_add_match(j, "PRIORITY=0", 0);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to add Emergency filter: %m");
+
+        r = sd_journal_seek_head(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to seek to start of jornal: %m");
+
+        r = sd_journal_next(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read next journal entry: %m");
+        if (r == 0) {
+                log_debug("No emergency level entries in the journal");
+                return r;
+        }
+
+        r = sd_journal_get_data(j, "MESSAGE", &d, &l);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        message = memdup_suffix0((const char*)d + STRLEN("MESSAGE="), l - STRLEN("MESSAGE="));
+        if (!message)
+                return log_oom();
+
+        *ret = TAKE_PTR(message);
+
+        return 0;
+}
+
+static int find_next_free_vt(int fd, int *ret_free_vt, int *ret_original_vt) {
+        size_t i;
+        struct vt_stat terminal_status;
+
+        assert(fd);
+        assert(ret_free_vt);
+        assert(ret_original_vt);
+
+        if (ioctl(fd, VT_GETSTATE, &terminal_status) < 0)
+                return -errno;
+
+        for (i = 0; i < sizeof(terminal_status.v_state) * 8; i++)
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+
+        if (i >= sizeof(terminal_status.v_state) * 8)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTTY), "No free VT found: %m");
+
+        *ret_free_vt = i;
+        *ret_original_vt = terminal_status.v_active;
+
+        return 0;
+}
+
+static int display_emergency_message_fullscreen(char *message) {
+        int r, free_vt = 0, original_vt = 0;
+        unsigned qr_code_start_row = 1, qr_code_start_column = 1;
+        char tty[STRLEN("/dev/tty") + DECIMAL_STR_MAX(int) + 1];
+        _cleanup_close_ int fd = -EBADF;
```

Comment:
@AdrianVovk 
doesn't this ```_cleanup_close_ ``` take care of such leakages?

Reply:
it does not, use close_and_replace()

Reply:
> it does not, use close_and_replace()

ohh...
I already pushed a patch where I separated the fds.

I use rather use close_and_replace()?

Reply:
> > it does not, use close_and_replace()
> 
> ohh... I already pushed a patch where I separated the fds.
> 
> I should rather use close_and_replace()?

Reply:
yes

Reply:
> yes

On it sir.